### PR TITLE
Added URL scheme checking and test functions

### PR DESCRIPTION
**Pull request by AdaephonBen**
This will fix issue #6397. 
Signed-off-by: AdaephonBen <ma18btech11011@iith.ac.in>

### CODE REVIEW

File: cmd/promtool/main_test.go [hunk 1]
```diff
@@ -23,6 +23,21 @@ import (
...
+	if urlWithScheme != "http://google.com" {
+		t.Errorf("unexpected value %s for urlWithScheme", urlWithScheme)
+	}
+	url = "http://prometheus.io"
+	urlWithScheme, err = addScheme(url)
+	testutil.Ok(t, err)
+	if urlWithScheme != "http://prometheus.io" {
+		t.Errorf("unexpected value %s for urlWithScheme", urlWithScheme)
+	}
+}
```
Comment (roidelapluie) [inline cmd/promtool/main_test.go:39]: this is a lot of repetition, can you make the test cases a []map[string]string?

e.g.

```
for input, result := range map[string]string{
 "google.com": "http://google.com",
} {
   ...
}
```

thanks

Comment (roidelapluie) [reply cmd/promtool/main_test.go:39]: Not that we also have testutl.Equals.

Comment (AdaephonBen) [reply cmd/promtool/main_test.go:36]: Thanks for pointing this out. I've fixed this in the latest commit.

File: cmd/promtool/main_test.go [hunk 2]
```diff
@@ -24,17 +24,16 @@ import (
...
-	if urlWithScheme != "http://prometheus.io" {
-		t.Errorf("unexpected value %s for urlWithScheme", urlWithScheme)
+	for input, result := range map[string]string{
+		"google.com":            "http://google.com",
+		"http://prometheus.io":  "http://prometheus.io",
+		"https://prometheus.io": "https://prometheus.io",
+	} {
+		urlWithScheme, err := addScheme(input)
+		testutil.Ok(t, err)
+		if urlWithScheme != result {
```
Comment (roidelapluie) [inline cmd/promtool/main_test.go]: testutil.Equals(t, expected, urlWithScheme)

File: cmd/promtool/main_test.go [hunk 3]
```diff
@@ -24,17 +24,16 @@ import (
...
-	testutil.Ok(t, err)
-	if urlWithScheme != "http://google.com" {
-		t.Errorf("unexpected value %s for urlWithScheme", urlWithScheme)
-	}
-	url = "http://prometheus.io"
-	urlWithScheme, err = addScheme(url)
-	testutil.Ok(t, err)
-	if urlWithScheme != "http://prometheus.io" {
-		t.Errorf("unexpected value %s for urlWithScheme", urlWithScheme)
+	for input, result := range map[string]string{
```
Comment (roidelapluie) [inline cmd/promtool/main_test.go:27]:
```suggestion
	for input, expected := range map[string]string{
```

**Review by roidelapluie — COMMENTED** [review]
It looks like we could turn queryRangeServer and queryInstantServer into URL's, and then we avoid the indirection to string. WDYT? 
e.g. querySeriesServer is already a URL.

You also miss debugPprofServer.

Thanks!

**Comment by AdaephonBen** [comment]
> It looks like we could turn queryRangeServer and queryInstantServer into URL's, and then we avoid the indirection to string. WDYT?
> e.g. querySeriesServer is already a URL.
> 
> You also miss debugPprofServer.
> 
> Thanks!

Yeah that makes sense. Will do it!
Do you mean that the `addScheme()` function also and return `*url.URL arguments` instead of strings?

**Comment by roidelapluie** [comment]
Yes, that'd be cleaner I think

**Comment by roidelapluie** [comment]
> Yes, that'd be cleaner I think

Also it should not return err anymore so you could inline it in ApiConfig ?

**Comment by AdaephonBen** [comment]
> Yes, that'd be cleaner I think

In this case, can I mutate the `*url.URL` argument directly and not have the `addScheme()` function return anything? Or is mutating it directly not a good idea?

**Comment by AdaephonBen** [comment]
> > Yes, that'd be cleaner I think
> 
> Also it should not return err anymore so you could inline it in Api Config ?

Cool, will do that.

**Comment by roidelapluie** [comment]
> > > Yes, that'd be cleaner I think
> > 
> > 
> > Also it should not return err anymore so you could inline it in Api Config ?
> 
> Cool, will do that.
> Edit: To do this, I'd have to open a PR to `client_golang` right? Should I do it before or after this PR has been merged?

No, it can all be done here

**Comment by AdaephonBen** [comment]
@roidelapluie I've made all the necessary changes.

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**